File: Go/Algorithms/mergeSort.go
package main

import "fmt"

func merge(left, right []int) []int {
	var result []int
	var i int = 0
	var j int = 0

	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	result = append(result, left[i:]...)
	result = append(result, right[j:]...)

	return result
}

func mergeSort(array []int) []int {
	if len(array) <= 1 {
		return array
	}
	var mid int = len(array) / 2
	var left []int = mergeSort(array[:mid])
	var right []int = mergeSort(array[mid:])

	return merge(left, right)
}

func main() {
	unsortedArray := []int{10, 9, 8, 7, 6, 5, 4, 3, 2, 1}
	fmt.Println(mergeSort(unsortedArray))
}

File: Nim/Algorithms/merge_sort.nim
proc merge(left, right: seq[int]): seq[int] = 
    var 
        arr: seq[int] = @[]
        i: int = 0
        j: int = 0
    
    while i < left.len and j < right.len:
        if left[i] < right[j]:
            arr.add(left[i])
            i += 1
        else:
            arr.add(right[j])
            j += 1
    
    arr.add(left[i ..< left.len])
    arr.add(right[j ..< right.len])

    return arr
    


proc merge_sort(arr: seq[int]): seq[int] =
    if arr.len <= 1:
        return arr

    let mid: int = arr.len div 2
    let left: seq[int] = merge_sort(arr[0 ..< mid])
    let right: seq[int] = merge_sort(arr[mid ..< arr.len])
    
    return merge(left, right)